### ISSUE TITLE
Create a sitemap

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#16508]
We'd like to expose an XML [sitemap](https://developers.google.com/search/docs/advanced/sitemaps/build-sitemap) to help search engines discover all listed content on AMO.

Several considerations need to be discussed:
- Performance. It likely needs to be pre-generated, or cached on disk.
- Refresh period, how fresh the contents needs to be
- Figuring out what pages go in it, how to deal with translations/languages in general

Django has [a framework for this](https://docs.djangoproject.com/en/3.0/ref/contrib/sitemaps/) we'd likely build upon. It supports paginating the sitemap and most basic things we'd need.

**Comment by eviljeff** [comment]
> cached on disk.

memcache okay?

**Comment by eviljeff** [comment]
> Figuring out what pages go in it, how to deal with translations/languages in general

This is the blocker.  

As a starter:
* /`<locale>`/firefox/ (i.e. the homepage)
* /`<locale>`/firefox/extensions/
* /`<locale>`/firefox/themes/
* /`<locale>`/firefox/language-tools/
* (the category pages at some point, but not now?)
* /`<locale>`/firefox/addon/`<addon.slug>`/ `for <addon> in Addon.objects.public()`

 `for <locale> in AMO_LANGUAGES` ?

+ mozilla collections maybe?

**Comment by eviljeff** [comment]
Although given most add-ons won't be localised in a significant number of locales (most not at all - only en-US) should we be generating (fake) localized add-on detail pages for every add-on?

**Comment by eviljeff** [comment]
And _all_ public addons?  Or just ones that are in a promoted group?

**Comment by diox** [comment]
>> cached on disk.
>
> memcache okay?

It's going to be a lot of data, probably not frequently accessed enough to stay in memcached. Which is why I was thinking of pre-generating, say once a day.

**Comment by diox** [comment]
Following discussion with @rraue: 

- We want every page to be included (excluding pages already excluded because of robots.txt: so, no search pages or non-mozilla collections).

- Start with `/en-US` only. The idea is that search engines will then find the pages in other languages easily through the `<link href="(...)" hreflang="(...)" rel="alternate">` links on every page. Need to double-check all pages included are going to have those links, though.

- Sitemap needs to be split so that it doesn't contain too many items in the same file. Django already has support for this.

- Sitemaps needs to be ordered by updated date - fresher content at the top.  (We can group sitemap files by "topic" to make this easier: have one with all base frontend pages, then others with all add-ons for instance)

- Worth mentioning again : it needs to be super fast for it to be worth it. It's ok for sitemaps to be out of date by 24 hours if that makes them faster to fetch.

**Comment by rraue** [comment]
Some basic thoughts:
- Include only urls that are 200 (no redirects, 404s etc. so https://addons.mozilla.org wouldnt be included, as its always redirected)
- Must not include any url with a meta or x-robot noindex
- Must not include any url blocked by robots.txt
- Must only include urls that are self-canonicalized (so no filter/parameter urls that point via canonical meta to another url)
- Must include all urls that aren't ruled out from one of the above (it's okay to start just with en-US but should include all even "unimportant" urls)
- Must include last modified

The sitemaps do not need to be reordered via the last modified if that's not something django does out of the box. If cached it will be not even a second for a crawler to get all urls and last-modified from the sitemap-index and all sitemaps included, no matter how many as its just text, no render.

**Comment by eviljeff** [comment]
#16904 will close this and add a minimal sitemap.xml, but there is more to do before I would consider the sitemap ready to be exposed to search engines. Copying what I wrote in the pr:

follow-ups needed for:  

- we don't have all the generic AMO pages - addons-server doesn't "know about" every frontend page so I just added the ones we already had defined in urls.pys (the vast majority of the urls in addons-server are behind auth, or apis)
- only the add-on detail pages are there. Other add-on specific pages need adding
- collections (I trivially added them, but dropped it as the patch was already big enough for now)
- localized urls and non-Firefox app urls
- job isn't in crontab.tpl yet so won't be run as a cron job
- Needs a patch to cloudops-deployment scripts to add sitemap.xml to ngnix (mozilla-services/cloudops-deployment#4246)
- locally frontend routing for urls without local&app is broken (mozilla/addons-frontend#10308)

**Comment by ioanarusiczki** [comment]
@eviljeff If this can be verified by qa please add some str.
Thanks!

**Comment by eviljeff** [comment]
> 
> 
> @eviljeff If this can be verified by qa please add some str.
> Thanks!

it needs the PR for https://github.com/mozilla-services/cloudops-deployment/issues/4246 (adding the nginx rule to pass `sitemap.xml` url to addons-server) merging by ops first.  I'll update when it's available.

**Comment by eviljeff** [comment]
the PR was merged, so `/sitemap.xml` is now resolved to addons-server.

The content that will be served is from a file; you can get a live generated version of any file (on addons-dev only) by putting `?debug`|`&debug` on the end of the url

**Comment by AlexandraMoga** [comment]
@eviljeff I can see that https://addons-dev.allizom.org/sitemap.xml is available on -dev and includes 4 sections: `?section=amo` - includes the pages served by the frontend, `?section=addons` and `?section=addons&p=2` - including listed add-ons published on AMO*, `?section=collections` - includes all collections created by mozilla.

*The `addons` sections are taking a few minutes to completely load all results (`addons` a bit more than `addons&p=2` since it returns 50k results)

> The content that will be served is from a file; you can get a live generated version of any file (on addons-dev only) by putting `?debug`|`&debug` on the end of the url

Does this mean that all the files/pages included in the sitemap sections mentioned above can be inspected by simply going to the page url like https://addons-dev.allizom.org/en-US/firefox/collections/4757633/webdeveloper/?debug|&debug for example?

**Comment by eviljeff** [comment]
> > The content that will be served is from a file; you can get a live generated version of any file (on addons-dev only) by putting `?debug`|`&debug` on the end of the url
> 
> Does this mean that all the files/pages included in the sitemap sections mentioned above can be inspected by simply going to the page url like [https://addons-dev.allizom.org/en-US/firefox/collections/4757633/webdeveloper/?debug|&debug](https://addons-dev.allizom.org/en-US/firefox/collections/4757633/webdeveloper/?debug%7C&debug) for example?

https://addons-dev.allizom.org/sitemap.xml?debug or
https://addons-dev.allizom.org/sitemap.xml?section=amo&debug

**Comment by eviljeff** [comment]
> *The `addons` sections are taking a few minutes to completely load all results (`addons` a bit more than `addons&p=2` since it returns 50k results)

Filed #16985 to try to make it faster
